File: src/main/java/me/wolfyscript/customcrafting/listeners/smithing/SmithingListener.java
/*
 *       ____ _  _ ____ ___ ____ _  _ ____ ____ ____ ____ ___ _ _  _ ____
 *       |    |  | [__   |  |  | |\/| |    |__/ |__| |___  |  | |\ | | __
 *       |___ |__| ___]  |  |__| |  | |___ |  \ |  | |     |  | | \| |__]
 *
 *       CustomCrafting Recipe creation and management tool for Minecraft
 *                      Copyright (C) 2021  WolfyScript
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package me.wolfyscript.customcrafting.listeners.smithing;

import com.wolfyscript.utilities.bukkit.world.items.reference.StackReference;
import me.wolfyscript.customcrafting.CustomCrafting;
import me.wolfyscript.customcrafting.recipes.CustomRecipeSmithing;
import me.wolfyscript.customcrafting.recipes.ICustomVanillaRecipe;
import me.wolfyscript.customcrafting.recipes.RecipeType;
import me.wolfyscript.customcrafting.recipes.data.RecipeData;
import me.wolfyscript.customcrafting.recipes.data.SmithingData;
import me.wolfyscript.customcrafting.recipes.items.Result;
import me.wolfyscript.customcrafting.recipes.items.target.MergeAdapter;
import me.wolfyscript.customcrafting.recipes.items.target.MergeOption;
import me.wolfyscript.utilities.api.inventory.custom_items.CustomItem;
import me.wolfyscript.utilities.util.NamespacedKey;
import me.wolfyscript.utilities.util.inventory.ItemUtils;
import me.wolfyscript.utilities.util.version.MinecraftVersion;
import me.wolfyscript.utilities.util.version.ServerVersion;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.Sound;
import org.bukkit.block.Block;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryAction;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.PrepareSmithingEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.SmithingInventory;
import org.bukkit.inventory.SmithingRecipe;

import java.util.HashMap;
import java.util.List;
import java.util.Objects;
import java.util.UUID;
import java.util.stream.Stream;

public class SmithingListener implements Listener {

    private final HashMap<UUID, SmithingData> preCraftedRecipes = new HashMap<>();
    private final CustomCrafting customCrafting;

    private static final boolean IS_1_20 = ServerVersion.isAfterOrEq(MinecraftVersion.of(1, 20, 0));

    public SmithingListener(CustomCrafting customCrafting) {
        this.customCrafting = customCrafting;
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    public void onPrepare(PrepareSmithingEvent event) {
        SmithingInventory inv = event.getInventory();
        var player = (Player) event.getView().getPlayer();

        if (!ItemUtils.isAirOrNull(event.getResult())) {
            if (Stream.of(inv.getStorageContents()).map(CustomItem::getByItemStack).anyMatch(i -> i != null && i.isBlockVanillaRecipes())
                    || Bukkit.getRecipesFor(event.getResult()).stream()
                    .anyMatch(recipe -> recipe instanceof SmithingRecipe smithingRecipe && customCrafting.getDisableRecipesHandler().getRecipes().contains(NamespacedKey.fromBukkit(smithingRecipe.getKey())))) {
                event.setResult(null);
            }
        }

        var template = IS_1_20 ? inv.getItem(0) : null;
        var base = inv.getItem(CustomRecipeSmithing.BASE_SLOT);
        var addition = inv.getItem(CustomRecipeSmithing.ADDITION_SLOT);
        preCraftedRecipes.put(player.getUniqueId(), null);

        customCrafting.getRegistries().getRecipes().get(RecipeType.SMITHING).stream()
                .sorted()
                .filter(recipe -> !recipe.isDisabled())
                .map(recipe -> recipe.check(player, event.getView(), template, base, addition))
                .filter(Objects::nonNull)
                .findFirst()
                .ifPresentOrElse(data -> {
                    preCraftedRecipes.put(player.getUniqueId(), data);
                    CustomRecipeSmithing recipe = data.getRecipe();

                    applyResult(event, inv, player, base, recipe.getResult(), recipe.isOnlyChangeMaterial(), recipe.getInternalMergeAdapters(), data);
                }, () -> {
                    if (ItemUtils.isAirOrNull(event.getResult())) return;
                    SmithingRecipe recipe = (SmithingRecipe) inv.getRecipe();
                    if (recipe == null || ICustomVanillaRecipe.isPlaceholderOrDisplayRecipe(recipe.getKey())) {
                        event.setResult(null);
                        customCrafting.getApi().getNmsUtil().getRecipeUtil().setCurrentRecipe(event.getView(), null);
                    }
                });
    }

    private static void applyResult(PrepareSmithingEvent event, SmithingInventory inv, Player player, ItemStack base, Result result, boolean onlyChangeMaterial, List<MergeAdapter> adapters, RecipeData<?> data) {
        //Process result
        Block block = inv.getLocation() != null ? inv.getLocation().getBlock() : null;
        StackReference chosenResult = result.item(player, block).orElse(StackReference.of(new ItemStack(Material.AIR)));
        ItemStack endResult = result.item(data, chosenResult, player, block);
        if (onlyChangeMaterial) {
            //Take the base item and just change the material.
            if (base == null) { // The base item may be null! Make sure to reset the result slot!
                event.setResult(null);
                return;
            }
            var baseCopy = base.clone();
            baseCopy.setType(endResult.getType());
            baseCopy.setAmount(endResult.getAmount());
            event.setResult(baseCopy);
        } else {
            if (!adapters.isEmpty()) {
                MergeOption option = new MergeOption(new int[]{CustomRecipeSmithing.BASE_SLOT});
                option.setAdapters(adapters);
                // This acts as if we appended extra merge adapters to the end of the result.
                // This makes it possible to implement the logic just once and not both in smithing listener and merge adapter.
                option.merge(data, player, block, chosenResult, endResult);
            }
            event.setResult(endResult);
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    public void onCollectResult(InventoryClickEvent event) {
        if (event.getClickedInventory() == null || !(event.getClickedInventory() instanceof SmithingInventory smithingInventory)) {
            return;
        }
        final var player = (Player) event.getWhoClicked();
        final var action = event.getAction();
        final var inventory = event.getClickedInventory();
        if (event.getSlot() == CustomRecipeSmithing.RESULT_SLOT && !ItemUtils.isAirOrNull(event.getCurrentItem())) {
            if (action == InventoryAction.PLACE_ALL || action == InventoryAction.PLACE_ONE || action == InventoryAction.PLACE_SOME || action == InventoryAction.SWAP_WITH_CURSOR) {
                return; // Ignore ingredient slots
            }
            if (preCraftedRecipes.get(player.getUniqueId()) == null) {
                // Vanilla Recipe, so lets stop here
                return;
            }

            final var smithingData = preCraftedRecipes.remove(player.getUniqueId());

            final var baseItem = inventory.getItem(CustomRecipeSmithing.BASE_SLOT) == null ? new ItemStack(Material.AIR) : inventory.getItem(CustomRecipeSmithing.BASE_SLOT).clone();
            final var additionItem = inventory.getItem(CustomRecipeSmithing.ADDITION_SLOT) == null ? new ItemStack(Material.AIR) : inventory.getItem(CustomRecipeSmithing.ADDITION_SLOT).clone();
            final var templateItem = inventory.getItem(0) == null ? new ItemStack(Material.AIR) : inventory.getItem(0).clone();

            ItemStack result = smithingInventory.getResult();
            if (result == null) {
                return;
            }

            event.setCancelled(true); // Cancel the event to prevent vanilla ingredient consumption
            // We only want to allow one craft. This is to prevent inconsistent behaviour.
            if (!handleInvalidRecipeClick(result, event)) {
                return;
            }

            player.playSound(player, Sound.BLOCK_SMITHING_TABLE_USE, 1.0F, 1.0F);
            smithingData.getResult().executeExtensions(inventory.getLocation() != null ? inventory.getLocation() : player.getLocation(), inventory.getLocation() != null, player);

            smithingData.template().ifPresent(reference -> inventory.setItem(0, reference.shrink(templateItem, 1, true, inventory, player, null)));
            inventory.setItem(CustomRecipeSmithing.BASE_SLOT, smithingData.base().map(reference -> reference.shrink(baseItem, 1, true, inventory, player, null)).orElse(baseItem));
            inventory.setItem(CustomRecipeSmithing.ADDITION_SLOT, smithingData.addition().map(reference -> reference.shrink(additionItem, 1, true, inventory, player, null)).orElse(additionItem));
        }
    }

    private boolean handleInvalidRecipeClick(ItemStack result, InventoryClickEvent event) {
        // Thanks to Spigot not allowing empty ingredients, we need our own way of collecting the result!
        // Otherwise, it would just cancel the click, because there is no valid recipe!
        if (event.getClick().isShiftClick()) {
            // No space in inventory! To not mess with dropping items, etc. we just cancel the click
            return event.getView().getBottomInventory().addItem(result).isEmpty();
        }
        // A quick implementation to collect the result. Things like moving the item to the hotbar won't work!
        if (ItemUtils.isAirOrNull(event.getCursor())) {
            Bukkit.getScheduler().runTask(customCrafting, () -> {
                event.getView().setCursor(result);
            });
        } else if (event.getCursor().isSimilar(result)) {
            if (event.getCursor().getAmount() + result.getAmount() > event.getCursor().getMaxStackSize()) {
                return false; // Again instead of doing weird stuff, just cancel
            }
            Bukkit.getScheduler().runTask(customCrafting, () -> {
                event.getView().getCursor().setAmount(event.getCursor().getAmount() + result.getAmount());
            });
        }
        return true;
    }

}

File: spigot/src/main/kotlin/com/wolfyscript/customcrafting/spigot/recipes/SmithingListener.kt
package com.wolfyscript.customcrafting.spigot.recipes

import com.github.benmanes.caffeine.cache.Caffeine
import com.wolfyscript.customcrafting.recipes.CustomRecipeSmithing
import com.wolfyscript.customcrafting.recipes.EvaluationContextImpl
import com.wolfyscript.customcrafting.recipes.RecipeTypes
import com.wolfyscript.customcrafting.recipes.SmithingUtils
import com.wolfyscript.customcrafting.recipes.data.RecipeEvaluationResult
import com.wolfyscript.customcrafting.recipes.data.RecipeInput
import com.wolfyscript.customcrafting.spigot.CustomCraftingSpigot
import com.wolfyscript.scafall.spigot.api.wrappers.utils.toPreciseGlobal
import com.wolfyscript.scafall.spigot.api.wrappers.utils.unwrap
import com.wolfyscript.scafall.spigot.api.wrappers.utils.wrap
import org.bukkit.Bukkit
import org.bukkit.Keyed
import org.bukkit.Material
import org.bukkit.Sound
import org.bukkit.entity.Player
import org.bukkit.event.Event
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryCloseEvent
import org.bukkit.event.inventory.PrepareSmithingEvent
import org.bukkit.event.inventory.SmithItemEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.SmithingInventory
import org.bukkit.persistence.PersistentDataType
import java.util.*
import kotlin.random.Random

class SmithingListener(val customCrafting: CustomCraftingSpigot) : Listener {

    val recipeCache = Caffeine.newBuilder().build<UUID, RecipeEvaluationResult<RecipeEvaluationResult.Data, CustomRecipeSmithing>>()
    val collectsResult = Caffeine.newBuilder().build<UUID, Boolean>()

    private fun getSmithingSeed(bukkitPlayer: Player): Long {
        var seed = bukkitPlayer.persistentDataContainer.get(
            CustomCraftingSpigot.playerSmithingSeedKey,
            PersistentDataType.LONG
        )
        if (seed == null) {
            seed = Random.Default.nextLong()
            bukkitPlayer.persistentDataContainer.set(
                CustomCraftingSpigot.playerSmithingSeedKey,
                PersistentDataType.LONG,
                seed
            )
        }
        return seed
    }

    @EventHandler
    fun onPrepare(event: PrepareSmithingEvent) {
        val inventory = event.inventory
        val resultStack = event.result

        recipeCache.invalidate(event.view.player.uniqueId)

        if (resultStack != null && resultStack.type != Material.AIR) {
            // Check for disabled vanilla recipes
            if (Bukkit.getRecipesFor(resultStack).any {
                    customCrafting.recipeManager.disabledRecipes.contains((it as Keyed).key.wrap())
                }) {
                event.result = null
            }
        }

        val context = EvaluationContextImpl((event.view.player as Player).wrap(), inventory.location?.toPreciseGlobal())
        val templateStack = inventory.getItem(0)
        val baseStack = inventory.getItem(1)
        val additionStack = inventory.getItem(2)

        val data = customCrafting.recipeManager.evaluateRecipesOfType(
            RecipeTypes.smithing.resolveOrThrow(),
            RecipeInput.SmithingRecipeInput.of(
                templateStack?.wrap(), baseStack?.wrap(), additionStack?.wrap()
            ),
            context
        )

        if (data != null) {
            val recipe = data.recipe.value ?: return

            recipeCache.put(event.view.player.uniqueId, data)

            val endResult = recipe.result.compute(
                data,
                context, Random(getSmithingSeed(event.view.player as Player))
            ).unwrap()

            if (baseStack == null) {
                event.result = null
                return
            }

            SmithingUtils.copyDataComponentsTo(baseStack.wrap(), endResult.wrap(), recipe.copyOptions!!)
            return
        }

        // No recipe was matched
        if (event.result == null || event.result!!.type == Material.AIR) {
            return
        }
        if (inventory.recipe == null || inventory.recipe!!.isPlaceholder() || inventory.recipe!!.isDisplay()) {
            event.result = null
        }

    }

    @EventHandler
    fun onCollectResult(event: SmithItemEvent) {
        val inventory = event.clickedInventory
        if (inventory == null || inventory !is SmithingInventory) {
            return
        }
        if (event.slot != 3 || event.currentItem == null || event.currentItem!!.type == Material.AIR) {
            return
        }

        val player = event.whoClicked as Player
        val action = event.action

        val data = recipeCache.getIfPresent(player.uniqueId) ?: return
        val recipe = data.recipe.value ?: return

        val resultStack = inventory.result
        if (resultStack == null || resultStack.type ==  Material.AIR) {
            return
        }

        event.result = Event.Result.DENY  // Cancel the event to prevent vanilla ingredient consumption

        if (event.isShiftClick) {
            if (event.view.bottomInventory.addItem(resultStack).isNotEmpty()) {
                return
            }
        }
        // A quick implementation to collect the result. Things like moving the item to the hotbar won't work!
        if (event.cursor.type == Material.AIR) {
            Bukkit.getScheduler().runTask(customCrafting.bootstrap.plugin, Runnable {
                event.view.setCursor(resultStack)
            })
        } else if (event.cursor.isSimilar(resultStack)) {
            if (event.cursor.amount + resultStack.amount > event.cursor.maxStackSize) {
                return // does not fit on cursor. cancel recipe
            }
            Bukkit.getScheduler().runTask(customCrafting.bootstrap.plugin, Runnable {
                event.view.cursor.amount = event.cursor.amount + resultStack.amount
            })
        }

        val context = EvaluationContextImpl((event.whoClicked as Player).wrap(), inventory.location?.toPreciseGlobal())

        recipe.result.runActions(context, 1)

        // TODO: craft remains
        // Use setContents so we batch slot updates and only cause one update, so PrepareSmithingEvent is just called once instead of three times.
        inventory.contents = arrayOf(
            data.data.bySlot(0)?.let {
                inventory.getItem(0)?.clone()?.apply {
                    amount -= it.matchedItemStackRef.amount
                } ?: ItemStack(Material.AIR)
            },
            data.data.bySlot(1)?.let {
                inventory.getItem(1)?.clone()?.apply {
                    amount -= it.matchedItemStackRef.amount
                } ?: ItemStack(Material.AIR)
            },
            data.data.bySlot(2)?.let {
                inventory.getItem(2)?.clone()?.apply {
                    amount -= it.matchedItemStackRef.amount
                } ?: ItemStack(Material.AIR)
            }
        )

        player.playSound(player, Sound.BLOCK_SMITHING_TABLE_USE, 1f, 1f)

        // Reset seed for next result generation
        player.persistentDataContainer.set(CustomCraftingSpigot.playerSmithingSeedKey, PersistentDataType.LONG, Random.Default.nextLong())
        recipeCache.invalidate(player.uniqueId)
    }

    @EventHandler
    fun onClose(event: InventoryCloseEvent) {
        if (event.inventory is SmithingInventory) {
            recipeCache.invalidate(event.player.uniqueId)
        }
    }

    @EventHandler
    fun onQuit(event: PlayerQuitEvent) {
        recipeCache.invalidate(event.player.uniqueId)
    }

}